### PR TITLE
Reduce repeated calls to django cache while calculating utm_source for listing endpoints

### PR DESCRIPTION
**Pull request by zawan-ila**
[PROD-3968](https://2u-internal.atlassian.net/browse/PROD-3968)

### CODE REVIEW

File: course_discovery/apps/api/cache.py [hunk 1]
```diff
@@ -165,3 +166,6 @@ def list(self, request, *args, **kwargs):
     )
     def retrieve(self, request, *args, **kwargs):
         return super().retrieve(request, *args, **kwargs)
+
+def get_utm_source_request_cache_key(partner, user):
+    return get_cache_key(partner=partner.id, user=user.id)
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/cache.py]: nit: missing end of file line.

File: course_discovery/apps/api/serializers.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 from django.utils.text import slugify
 from django.utils.translation import gettext_lazy as _
 from django_countries.serializer_fields import CountryField
+from edx_django_utils.cache import RequestCache, get_cache_key
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: These are unused.

File: course_discovery/apps/api/utils.py [hunk 3]
```diff
@@ -352,3 +353,20 @@ def push_to_studio(self, course_run, create=False, old_course_run_key=None, user
             self.create_course_run_in_studio(course_run, user=user)
         else:
             self.update_course_run_details_in_studio(course_run)
+
+def use_request_cache(cache_name, key_func):
+    def inner(fn):
+        @functools.wraps(fn)
+        def foo(*args, **kwargs):
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/utils.py]: nit: foo is not a good name here.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py [hunk 4]
```diff
@@ -352,6 +353,26 @@ def test_list(self):
...
+        self.partner.lms_url = "http://localhost:8000"
+        self.partner.save()
+        with mock.patch('course_discovery.apps.core.api_client.lms.LMSAPIClient.get_api_access_request', return_value={
+                "company_name": "Example Inc",
+                "company_address": "Example Address",
+        }) as mocked_api_access_request:
+            with override_switch('use_company_name_as_utm_source_value', True):
+                url = reverse('api:v1:course-list')
+                self.client.get(url)
+                mocked_api_access_request.assert_called_once()
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_courses.py:375]: nit: should we not be checking utm call's count here too?

**Review by DawoudSheraz — APPROVED** [review]

File: course_discovery/apps/api/utils.py [hunk 5]
```diff
@@ -352,3 +353,20 @@ def push_to_studio(self, course_run, create=False, old_course_run_key=None, user
             self.create_course_run_in_studio(course_run, user=user)
         else:
             self.update_course_run_details_in_studio(course_run)
+
+
+def use_request_cache(cache_name, key_func):
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/api/utils.py:358]: A docstring explaining it's usage would be nice.

File: course_discovery/apps/api/utils.py [hunk 6]
```diff
@@ -352,3 +353,20 @@ def push_to_studio(self, course_run, create=False, old_course_run_key=None, user
...
+        def wrapped(*args, **kwargs):
+            cache = RequestCache(cache_name)
+            cache_key = key_func(*args, **kwargs)
+            cached_response = cache.get_cached_response(cache_key)
+            if cached_response.is_found:
+                return cached_response.value
+
+            ret_val = fn(*args, **kwargs)
+            cache.set(cache_key, ret_val)
+            return ret_val
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/api/utils.py]: Naming and readability can be improved like:
Rename `inner` to `decorator` to make it clear that this is a decorator function.
Rename `wrapped` to `wrapper` to follow common convention in Python decorator patterns.
Rename `ret_val` to something like `result`

Comment (Ali-D-Akbar) [reply course_discovery/apps/api/utils.py]: A bunch of lines here n there to improve readability.

**Review by Ali-D-Akbar — APPROVED** [review]
Functionally looks good. Just a bunch of cosmetic suggestions.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py (hunk 4 above)
Comment (zawan-ila) [reply course_discovery/apps/api/v1/tests/test_views/test_courses.py:375]: The `get_utm_source_for_user` function will still be called as many times as it is being currently called. (In this case twice). However, the stuff inside of it (which includes `get_api_access_request`) will only be called once. And any subsequent invocations will fetch the result from the RequestCache

File: course_discovery/apps/api/utils.py (hunk 6 above)
Comment (zawan-ila) [reply course_discovery/apps/api/utils.py]: [curious] Was the first of these comments generated through some AI tool?

Comment (Ali-D-Akbar) [reply course_discovery/apps/api/utils.py]: I wrote this myself although I got the naming conventions from GPT :D

Comment (zawan-ila) [reply course_discovery/apps/api/utils.py]: I have renamed `wrapped` and `ret_val` but not `inner`. I feel it is better as is.

File: course_discovery/apps/api/utils.py (hunk 5 above)
Comment (zawan-ila) [reply course_discovery/apps/api/utils.py:358]: Done. Let me know if it is helpful or needs improvement.

**Pull request merged**